routing/cross_mwm_index_graph.cpp

```diff
@@ -3,30 +3,69 @@
 #include "platform/country_file.hpp"
 
 #include "base/macros.hpp"
+#include "base/stl_helpers.hpp"
 
+#include <algorithm>
 #include <utility>
 
+using namespace platform;
+using namespace routing;
+using namespace std;
+
 namespace
 {
-void FillTransitionSegments(routing::OsrmFtSegMapping const & segMapping, routing::TWrittenNodeId nodeId,
-                            routing::NumMwmId mwmId, std::set<routing::Segment> & transitionSegments)
+bool GetTransitionSegment(OsrmFtSegMapping const & segMapping, TWrittenNodeId nodeId,
+                          NumMwmId numMwmId, Segment & segment)
 {
   auto const range = segMapping.GetSegmentsRange(nodeId);
   for (size_t segmentIndex = range.first; segmentIndex != range.second; ++segmentIndex)
   {
-    routing::OsrmMappingTypes::FtSeg seg;
+    OsrmMappingTypes::FtSeg seg;
     // The meaning of node id in osrm is an edge between two joints.
     // So, it's possible to consider the first valid segment from the range which returns by GetSegmentsRange().
     segMapping.GetSegmentByIndex(segmentIndex, seg);
     if (!seg.IsValid())
       continue;
 
     CHECK_NOT_EQUAL(seg.m_pointStart, seg.m_pointEnd, ());
-    transitionSegments.emplace(seg.m_fid, min(seg.m_pointStart, seg.m_pointEnd), mwmId, seg.IsForward());
-    return;
+    segment = Segment(seg.m_fid, min(seg.m_pointStart, seg.m_pointEnd), numMwmId, seg.IsForward());
+    return true;
   }
   LOG(LERROR, ("No valid segments in the range returned by OsrmFtSegMapping::GetSegmentsRange(", nodeId,
-               "). Num mwm id:", mwmId));
+               "). Num mwm id:", numMwmId));
+  return false;
+}
+
+void AddTransitionSegment(OsrmFtSegMapping const & segMapping, TWrittenNodeId nodeId,
+                          NumMwmId numMwmId, std::vector<Segment> & segments)
+{
+  Segment key;
+  if (GetTransitionSegment(segMapping, nodeId, numMwmId, key))
+    segments.push_back(move(key));
+}
+
+void FillTransitionSegments(OsrmFtSegMapping const & segMapping, TWrittenNodeId nodeId,
+                            NumMwmId numMwmId, ms::LatLon const & latLon,
+                            std::map<Segment, ms::LatLon> & transitionSegments)
+{
+  Segment key;
+  if (!GetTransitionSegment(segMapping, nodeId, numMwmId, key))
+    return;
+
+  auto const p = transitionSegments.emplace(move(key), latLon);
+  if (p.second)
+    return;
+
+  // @TODO(bykoianko) It's necessary to investigate this case.
+  LOG(LWARNING, ("Trying to insert an inserted transition segment. The key:",
+                 *p.first, ". Inserted value", p.second, "Trying to insert value:", latLon));
```

Comment:
1. I change the message to
```
  LOG(LWARNING, ("Trying to insert an inserted transition segment. The key:",
                 *p.first, "Trying to insert value:", latLon));
```

2.
> Do we really need such generic message?

The problem is I still don't fully understand the reasons of such behavior. The most likely there are In(Out)goingCrossNode(s) with the same Node Ids in mappingPtr->m_crossContext.m_in(out)goingIndex. But I don't know for sure. Now I left a todo and a warning here.